File: tests/DesktopGrass.Win2D.Tests/SimTests/CatTests.cs
// CatTests.cs - §17 Cat critter subsystem.
//
// Mirrors tests/DesktopGrass.Native.Tests/src/cat_tests.cpp.

using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class CatTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED,
                                double monitorWidth = Monitor1920,
                                double density = 1.0)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, monitorWidth, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, monitorWidth);
        sim.ResetEntities(seed);
        return sim;
    }

    private static int CountKind(Sim sim, EntityKind kind) =>
        sim.Entities.Count(e => e.Kind == kind);

    private static int IndexOfFirstKind(Sim sim, EntityKind kind)
    {
        for (int i = 0; i < sim.Entities.Count; i++)
            if (sim.Entities[i].Kind == kind) return i;
        return -1;
    }

    private static void KeepFirstCatOnly(Sim sim)
    {
        int idx = IndexOfFirstKind(sim, EntityKind.Cat);
        Assert.True(idx >= 0);
        var cat = sim.Entities[idx];
        sim.Entities.Clear();
        sim.Entities.Add(cat);
    }

    [Fact]
    public void CritterKindCatAndCritterCountArePinned()
    {
        Assert.Equal(0, (int)CritterKind.None);
        Assert.Equal(1, (int)CritterKind.Sheep);
        Assert.Equal(2, (int)CritterKind.Cat);
        Assert.Equal(3, (int)CritterKind.Bunny);
        Assert.Equal(4, Constants.CRITTER_COUNT);
        Assert.Equal(CritterKind.None, Constants.CRITTER_DEFAULT);
    }

    [Fact]
    public void EntityKindCatIsPinned()
    {
        Assert.Equal(0, (int)EntityKind.None);
        Assert.Equal(1, (int)EntityKind.Tumbleweed);
        Assert.Equal(2, (int)EntityKind.Snowflake);
        Assert.Equal(3, (int)EntityKind.Sheep);
        Assert.Equal(4, (int)EntityKind.Cat);
    }

    [Fact]
    public void CatConstantsArePinnedToSpecValues()
    {
        Assert.Equal(1, Constants.CAT_COUNT_MIN);
        Assert.Equal(2, Constants.CAT_COUNT_MAX);
        Assert.Equal(10.0, Constants.CAT_WALK_SPEED_MIN);
        Assert.Equal(22.0, Constants.CAT_WALK_SPEED_MAX);
        Assert.Equal(60.0, Constants.CAT_POUNCE_SPEED);

        Assert.Equal(11.0, Constants.CAT_BODY_RADIUS);
        Assert.Equal(7.0, Constants.CAT_BODY_HEIGHT);
        Assert.Equal(4.5, Constants.CAT_HEAD_RADIUS);
        Assert.Equal(5.0, Constants.CAT_LEG_LENGTH);
        Assert.Equal(13.0, Constants.CAT_TAIL_LENGTH);
        Assert.Equal(1.6, Constants.CAT_TAIL_THICKNESS);
        Assert.Equal(4.5, Constants.CAT_EAR_HEIGHT);

        Assert.Equal(0xFF6B6259u, Constants.CAT_BODY_COLOR);
        Assert.Equal(0xFF3D3733u, Constants.CAT_LEG_COLOR);
        Assert.Equal(0xFF6B6259u, Constants.CAT_FACE_COLOR);
        Assert.Equal(0xFF3D3733u, Constants.CAT_EAR_COLOR);
        Assert.Equal(0xFF1A1614u, Constants.CAT_INK_COLOR);

        Assert.Equal(0.50, Constants.CAT_WALK_PERIOD);
        Assert.Equal(1.6, Constants.CAT_LEG_CYCLE_AMP);
        Assert.Equal(0.4, Constants.CAT_HEAD_BOB_AMP);
        Assert.Equal(1.2, Constants.CAT_TAIL_SWAY_FREQ);
        Assert.Equal(0.35, Constants.CAT_TAIL_SWAY_AMP);

        Assert.Equal(Constants.SHEEP_STATE_WALKING, Constants.CAT_STATE_WALKING);
        Assert.Equal(Constants.SHEEP_STATE_IDLE, Constants.CAT_STATE_IDLE);
        Assert.Equal(Constants.SHEEP_STATE_SLEEPING, Constants.CAT_STATE_SLEEPING);
        Assert.Equal(Constants.SHEEP_STATE_HOPPING, Constants.CAT_STATE_POUNCING);

        Assert.Equal(6.0, Constants.CAT_WALK_DURATION_MIN);
        Assert.Equal(10.0, Constants.CAT_WALK_DURATION_MAX);
        Assert.Equal(4.0, Constants.CAT_IDLE_DURATION_MIN);
        Assert.Equal(8.0, Constants.CAT_IDLE_DURATION_MAX);
        Assert.Equal(20.0, Constants.CAT_SLEEP_DURATION_MIN);
        Assert.Equal(40.0, Constants.CAT_SLEEP_DURATION_MAX);
        Assert.Equal(0.45, Constants.CAT_POUNCE_DURATION);

        Assert.Equal(0.65, Constants.CAT_IDLE_PROBABILITY);
        Assert.Equal(0.30, Constants.CAT_SLEEP_PROBABILITY);
        Assert.Equal(0.50, Constants.CAT_SLEEP_FROM_IDLE_PROB_DEFAULT);
        Assert.Equal(0.20, Constants.CAT_SLEEP_FROM_IDLE_PROB_MORNING);
        Assert.Equal(0.85, Constants.CAT_SLEEP_FROM_IDLE_PROB_NIGHT);

        Assert.Equal(80.0, Constants.CAT_POUNCE_RADIUS);
        Assert.Equal(9.0, Constants.CAT_POUNCE_HEIGHT);
        Assert.Equal(100.0, Constants.CAT_CURIOUS_RADIUS);
        Assert.Equal(0.7, Constants.CAT_CURIOUS_HEAD_TURN_MAX);
    }

    [Fact]
    public void SimDefaultsToNoneAndDoesNotGenerateCatsUntilSelected()
    {
        var sim = BuildSim();
        Assert.Equal(CritterKind.None, sim.CurrentCritter);
        Assert.Equal(0, CountKind(sim, EntityKind.Cat));

        sim.SetCritter(CritterKind.Cat);
        Assert.True(CountKind(sim, EntityKind.Cat) >= Constants.CAT_COUNT_MIN);
    }

    [Fact]
    public void SetCritterCatProducesDeterministicCats()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);

        Assert.Equal(CritterKind.Cat, sim.CurrentCritter);
        int k = CountKind(sim, EntityKind.Cat);
        Assert.InRange(k, Constants.CAT_COUNT_MIN, Constants.CAT_COUNT_MAX);

        foreach (var e in sim.Entities)
        {
            if (e.Kind != EntityKind.Cat) continue;
            Assert.Equal(Constants.CAT_STATE_WALKING, e.State);
            Assert.InRange(e.StateTimer,
                Constants.CAT_WALK_DURATION_MIN,
                Constants.CAT_WALK_DURATION_MAX);
            double speed = System.Math.Abs(e.Vx);
            Assert.InRange(speed,
                Constants.CAT_WALK_SPEED_MIN,
                Constants.CAT_WALK_SPEED_MAX);
            double margin = e.Size + 8.0;
            Assert.InRange(e.X, margin, sim.MonitorWidth - margin);
            Assert.Equal(sim.WindowHeight - Constants.CAT_BODY_HEIGHT - Constants.CAT_LEG_LENGTH,
                         e.Y, 9);
            Assert.Equal(Constants.CAT_BODY_RADIUS, e.Size, 9);
            Assert.True(e.Lifetime < 0.0);
            Assert.True(e.NameIndex < Constants.CAT_NAME_POOL.Length);
            Assert.True(e.CoatVariantIndex < Constants.CAT_COAT_VARIANT_COUNT);
        }
    }

    [Fact]
    public void CatPrngDrawOrderMatchesSideStream()
    {
        // count, then per-cat: x, speed, dir-coin, seed, stateTimer, nameIndex, coatVariantIndex
        var side = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.CRITTER_PRNG_SALT);

        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);

        double countDraw = side.Uniform(Constants.CAT_COUNT_MIN, Constants.CAT_COUNT_MAX + 1);
        int expectedCount = (int)System.Math.Floor(countDraw);
        if (expectedCount < Constants.CAT_COUNT_MIN) expectedCount = Constants.CAT_COUNT_MIN;
        if (expectedCount > Constants.CAT_COUNT_MAX) expectedCount = Constants.CAT_COUNT_MAX;
        Assert.Equal(expectedCount, CountKind(sim, EntityKind.Cat));

        int seen = 0;
        foreach (var e in sim.Entities)
        {
            if (e.Kind != EntityKind.Cat) continue;
            double margin = Constants.CAT_BODY_RADIUS + 8.0;
            double expectedX = side.Uniform(margin, Monitor1920 - margin);
            double expectedSpeed = side.Uniform(Constants.CAT_WALK_SPEED_MIN, Constants.CAT_WALK_SPEED_MAX);
            double dirCoin = side.Uniform(0.0, 1.0);
            double expectedDir = dirCoin < 0.5 ? -1.0 : 1.0;
            uint expectedSeed = side.NextU32();
            double expectedTimer = side.Uniform(Constants.CAT_WALK_DURATION_MIN,
                                                Constants.CAT_WALK_DURATION_MAX);
            byte expectedNameIndex = (byte)side.Index((uint)Constants.CAT_NAME_POOL.Length);
            byte expectedCoatVariantIndex = (byte)side.Index((uint)Constants.CAT_COAT_VARIANT_COUNT);

            Assert.Equal(expectedX, e.X, 9);
            Assert.Equal(expectedSpeed * expectedDir, e.Vx, 9);
            Assert.Equal(expectedSeed, e.Seed);
            Assert.Equal(expectedTimer, e.StateTimer, 9);
            Assert.Equal(expectedNameIndex, e.NameIndex);
            Assert.Equal(expectedCoatVariantIndex, e.CoatVariantIndex);
            seen++;
        }
        Assert.Equal(expectedCount, seen);
    }

    [Fact]
    public void SetCritterNoneClearsAmbientCats()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        Assert.True(CountKind(sim, EntityKind.Cat) >= Constants.CAT_COUNT_MIN);

        sim.SetCritter(CritterKind.None);
        Assert.Equal(CritterKind.None, sim.CurrentCritter);
        Assert.Equal(0, CountKind(sim, EntityKind.Cat));
        Assert.Equal(0, CountKind(sim, EntityKind.Bunny));
    }

    [Fact]
    public void SwitchingBetweenCritterSpeciesReplacesPriorSpecies()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        Assert.True(CountKind(sim, EntityKind.Cat) >= Constants.CAT_COUNT_MIN);
        Assert.Equal(0, CountKind(sim, EntityKind.Sheep));

        sim.SetCritter(CritterKind.Sheep);
        Assert.Equal(0, CountKind(sim, EntityKind.Cat));
        Assert.True(CountKind(sim, EntityKind.Sheep) >= Constants.SHEEP_COUNT_MIN);

        sim.SetCritter(CritterKind.Cat);
        Assert.Equal(0, CountKind(sim, EntityKind.Sheep));
        Assert.True(CountKind(sim, EntityKind.Cat) >= Constants.CAT_COUNT_MIN);
    }

    [Fact]
    public void SetSceneGatesActiveCatToGrass()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        int catsGrass = CountKind(sim, EntityKind.Cat);
        Assert.True(catsGrass >= Constants.CAT_COUNT_MIN);

        sim.SetScene(Scene.Desert);
        Assert.Equal(CritterKind.Cat, sim.CurrentCritter);
        Assert.Equal(0, CountKind(sim, EntityKind.Cat));

        sim.SetScene(Scene.Winter);
        Assert.Equal(CritterKind.Cat, sim.CurrentCritter);
        Assert.Equal(0, CountKind(sim, EntityKind.Cat));

        sim.SetScene(Scene.Grass);
        Assert.Equal(catsGrass, CountKind(sim, EntityKind.Cat));
    }

    [Fact]
    public void ClickWithinCatPounceRadiusPouncesTowardClick()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        KeepFirstCatOnly(sim);

        var cat = sim.Entities[0];
        cat.X = 500.0;
        cat.Vx = -Constants.CAT_WALK_SPEED_MIN;
        cat.Age = 5.0;
        sim.Entities[0] = cat;

        sim.ApplyClick(cat.X + 16.0, sim.WindowHeight - 20.0, 0.0);

        var after = sim.Entities[0];
        Assert.Equal(Constants.CAT_STATE_POUNCING, after.State);
        Assert.Equal(Constants.CAT_POUNCE_DURATION, after.StateTimer, 9);
        Assert.Equal(0.0, after.Age, 9);
        Assert.Equal(Constants.CAT_POUNCE_SPEED, after.Vx, 9);
    }

    [Fact]
    public void ClickOutsideCatPounceRadiusLeavesCatAlone()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        KeepFirstCatOnly(sim);

        var cat = sim.Entities[0];
        cat.X = 500.0;
        cat.Vx = -Constants.CAT_WALK_SPEED_MIN;
        sim.Entities[0] = cat;
        byte stateBefore = cat.State;
        double vxBefore = cat.Vx;

        sim.ApplyClick(cat.X + Constants.CAT_POUNCE_RADIUS + 5.0,
                       sim.WindowHeight - 20.0, 0.0);

        var after = sim.Entities[0];
        Assert.Equal(stateBefore, after.State);
        Assert.Equal(vxBefore, after.Vx, 9);
    }

    [Fact]
    public void CatsDoNotGreetOtherCats()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        KeepFirstCatOnly(sim);

        var first = sim.Entities[0];
        first.X = 400.0;
        first.Vx = Constants.CAT_WALK_SPEED_MIN;
        first.State = Constants.CAT_STATE_WALKING;
        first.StateTimer = 10.0;
        first.Age = Constants.SHEEP_GREET_MIN_AGE + 1.0;
        var second = first;
        second.X = first.X + 20.0;
        second.Vx = -Constants.CAT_WALK_SPEED_MIN;
        sim.Entities.Clear();
        sim.Entities.Add(first);
        sim.Entities.Add(second);

        sim.TickEntities(0.016);

        Assert.Equal(2, CountKind(sim, EntityKind.Cat));
        Assert.All(sim.Entities.Where(e => e.Kind == EntityKind.Cat),
            e => Assert.NotEqual(Constants.SHEEP_STATE_GREETING, e.State));
    }

    [Fact]
    public void CatsDoNotGreetSheep()
    {
        var sim = BuildSim();
        sim.SetCritter(CritterKind.Cat);
        KeepFirstCatOnly(sim);

        var cat = sim.Entities[0];
        cat.X = 400.0;
        cat.Vx = Constants.CAT_WALK_SPEED_MIN;
        cat.State = Constants.CAT_STATE_WALKING;
        cat.StateTimer = 10.0;
        cat.Age = Constants.SHEEP_GREET_MIN_AGE + 1.0;

        var sheep = new Entity
        {
            Kind = EntityKind.Sheep,
            Size = Constants.SHEEP_BODY_RADIUS,
            X = cat.X + 20.0,
            Y = sim.WindowHeight - Constants.SHEEP_BODY_HEIGHT - Constants.SHEEP_LEG_LENGTH,
            Vx = -Constants.SHEEP_WALK_SPEED_MIN,
            Age = Constants.SHEEP_GREET_MIN_AGE + 1.0,
            Lifetime = -1.0,
            State = Constants.SHEEP_STATE_WALKING,
            StateTimer = 10.0,
        };

        sim.Entities.Clear();
        sim.Entities.Add(cat);
        sim.Entities.Add(sheep);

        sim.TickEntities(0.016);

        Assert.Equal(1, CountKind(sim, EntityKind.Cat));
        Assert.Equal(1, CountKind(sim, EntityKind.Sheep));
        Assert.All(sim.Entities, e => Assert.NotEqual(Constants.SHEEP_STATE_GREETING, e.State));
    }

    [Fact]
    public void CatTimeOfDaySleepBiasIsPinned()
    {
        Assert.Equal(0.85, Sim.CatSleepProbForLocalHour(2));
        Assert.Equal(0.20, Sim.CatSleepProbForLocalHour(8));
        Assert.Equal(0.50, Sim.CatSleepProbForLocalHour(15));
        Assert.Equal(0.85, Sim.CatSleepProbForLocalHour(23));
    }
}

File: tests/DesktopGrass.Native.Tests/src/cat_tests.cpp
// cat_tests.cpp
//
// §17 Cat critter tests. Mirrors Win2D CatTests.cs.

#include "TestHelpers.h"
#include "Sim.h"

#include <algorithm>
#include <cmath>

using namespace desktopgrass;

namespace {

int count_kind(const Sim& sim, EntityKind kind) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [kind](const Entity& e) { return e.kind == kind; }));
}

Entity* first_kind(Sim& sim, EntityKind kind) {
    for (Entity& e : sim.entities) if (e.kind == kind) return &e;
    return nullptr;
}

const Entity* first_kind(const Sim& sim, EntityKind kind) {
    for (const Entity& e : sim.entities) if (e.kind == kind) return &e;
    return nullptr;
}

void keep_first_cat_only(Sim& sim) {
    Entity* cat = first_kind(sim, EntityKind::Cat);
    Assert::IsTrue(cat != nullptr);
    const Entity copy = *cat;
    sim.entities.clear();
    sim.entities.push_back(copy);
}

InputEvent click_event(double x, double y) {
    InputEvent ev{};
    ev.type = EventType::Click;
    ev.x = x;
    ev.y = y;
    ev.time = 0.0;
    return ev;
}

} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(CatTests)
{
public:
TEST_METHOD(CritterKindCatAndCRITTERCOUNTArePinned) {
    Assert::IsTrue(static_cast<int>(CritterKind::None)  == 0);
    Assert::IsTrue(static_cast<int>(CritterKind::Sheep) == 1);
    Assert::IsTrue(static_cast<int>(CritterKind::Cat)   == 2);
    Assert::IsTrue(static_cast<int>(CritterKind::Bunny) == 3);
    Assert::IsTrue(CRITTER_COUNT == 4);
    Assert::IsTrue(CRITTER_DEFAULT == CritterKind::None);
}

TEST_METHOD(EntityKindCatIsPinned) {
    Assert::IsTrue(static_cast<int>(EntityKind::None)       == 0);
    Assert::IsTrue(static_cast<int>(EntityKind::Tumbleweed) == 1);
    Assert::IsTrue(static_cast<int>(EntityKind::Snowflake)  == 2);
    Assert::IsTrue(static_cast<int>(EntityKind::Sheep)      == 3);
    Assert::IsTrue(static_cast<int>(EntityKind::Cat)        == 4);
}

TEST_METHOD(CatConstantsArePinnedToSpecValues) {
    Assert::IsTrue(CAT_COUNT_MIN == 1);
    Assert::IsTrue(CAT_COUNT_MAX == 2);
    Assert::IsTrue(CAT_WALK_SPEED_MIN == Near(10.0));
    Assert::IsTrue(CAT_WALK_SPEED_MAX == Near(22.0));
    Assert::IsTrue(CAT_POUNCE_SPEED   == Near(60.0));

    Assert::IsTrue(CAT_BODY_RADIUS    == Near(11.0));
    Assert::IsTrue(CAT_BODY_HEIGHT    == Near(7.0));
    Assert::IsTrue(CAT_HEAD_RADIUS    == Near(4.5));
    Assert::IsTrue(CAT_LEG_LENGTH     == Near(5.0));
    Assert::IsTrue(CAT_TAIL_LENGTH    == Near(13.0));
    Assert::IsTrue(CAT_TAIL_THICKNESS == Near(1.6));
    Assert::IsTrue(CAT_EAR_HEIGHT     == Near(4.5));

    Assert::IsTrue(CAT_BODY_COLOR == 0xFF6B6259u);
    Assert::IsTrue(CAT_LEG_COLOR  == 0xFF3D3733u);
    Assert::IsTrue(CAT_FACE_COLOR == 0xFF6B6259u);
    Assert::IsTrue(CAT_EAR_COLOR  == 0xFF3D3733u);
    Assert::IsTrue(CAT_INK_COLOR  == 0xFF1A1614u);

    Assert::IsTrue(CAT_WALK_PERIOD    == Near(0.50));
    Assert::IsTrue(CAT_LEG_CYCLE_AMP  == Near(1.6));
    Assert::IsTrue(CAT_HEAD_BOB_AMP   == Near(0.4));
    Assert::IsTrue(CAT_TAIL_SWAY_FREQ == Near(1.2));
    Assert::IsTrue(CAT_TAIL_SWAY_AMP  == Near(0.35));

    Assert::IsTrue(CAT_STATE_WALKING  == SHEEP_STATE_WALKING);
    Assert::IsTrue(CAT_STATE_IDLE     == SHEEP_STATE_IDLE);
    Assert::IsTrue(CAT_STATE_SLEEPING == SHEEP_STATE_SLEEPING);
    Assert::IsTrue(CAT_STATE_POUNCING == SHEEP_STATE_HOPPING);

    Assert::IsTrue(CAT_WALK_DURATION_MIN  == Near(6.0));
    Assert::IsTrue(CAT_WALK_DURATION_MAX  == Near(10.0));
    Assert::IsTrue(CAT_IDLE_DURATION_MIN  == Near(4.0));
    Assert::IsTrue(CAT_IDLE_DURATION_MAX  == Near(8.0));
    Assert::IsTrue(CAT_SLEEP_DURATION_MIN == Near(20.0));
    Assert::IsTrue(CAT_SLEEP_DURATION_MAX == Near(40.0));
    Assert::IsTrue(CAT_POUNCE_DURATION    == Near(0.45));

    Assert::IsTrue(CAT_IDLE_PROBABILITY == Near(0.65));
    Assert::IsTrue(CAT_SLEEP_PROBABILITY == Near(0.30));
    Assert::IsTrue(CAT_SLEEP_FROM_IDLE_PROB == Near(0.50));

    Assert::IsTrue(CAT_POUNCE_RADIUS == Near(80.0));
    Assert::IsTrue(CAT_POUNCE_HEIGHT == Near(9.0));
    Assert::IsTrue(CAT_CURIOUS_RADIUS == Near(100.0));
    Assert::IsTrue(CAT_CURIOUS_HEAD_TURN_MAX == Near(0.7));
}

TEST_METHOD(SimInitDefaultsToNoneAndDoesNotGenerateCatsUntilSelected) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    Assert::IsTrue(sim.currentCritter == CritterKind::None);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 0);

    sim_set_critter(sim, CritterKind::Cat);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) >= CAT_COUNT_MIN);
}

TEST_METHOD(SimSetCritterCatProducesDeterministicCats) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);

    Assert::IsTrue(sim.currentCritter == CritterKind::Cat);
    const int k = count_kind(sim, EntityKind::Cat);
    Assert::IsTrue(k >= CAT_COUNT_MIN);
    Assert::IsTrue(k <= CAT_COUNT_MAX);

    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Cat) continue;
        Assert::IsTrue(e.state == CAT_STATE_WALKING);
        Assert::IsTrue(e.stateTimer >= CAT_WALK_DURATION_MIN);
        Assert::IsTrue(e.stateTimer <  CAT_WALK_DURATION_MAX);
        Assert::IsTrue(std::fabs(e.vx) >= CAT_WALK_SPEED_MIN);
        Assert::IsTrue(std::fabs(e.vx) <  CAT_WALK_SPEED_MAX);
        const double margin = e.size + 8.0;
        Assert::IsTrue(e.x >= margin);
        Assert::IsTrue(e.x <= sim.monitorWidth - margin);
        Assert::IsTrue(e.y == Near(sim.windowHeight - CAT_BODY_HEIGHT - CAT_LEG_LENGTH));
        Assert::IsTrue(e.size == Near(CAT_BODY_RADIUS));
        Assert::IsTrue(e.lifetime < 0.0);
        Assert::IsTrue(e.nameIndex < sizeof(CAT_NAME_POOL) / sizeof(CAT_NAME_POOL[0]));
        Assert::IsTrue(e.coatVariantIndex < CAT_COAT_VARIANT_COUNT);
    }
}

TEST_METHOD(CatPRNGDrawOrderMatchesASideStream) {
    // count, then per-cat: x, speed, dir-coin, seed, stateTimer, nameIndex, coatVariantIndex
    Prng side;
    prng_init(side, CANONICAL_TEST_SEED ^ CRITTER_PRNG_SALT);

    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);

    const double countDraw = prng_uniform(side, CAT_COUNT_MIN, CAT_COUNT_MAX + 1);
    int expectedCount = static_cast<int>(std::floor(countDraw));
    if (expectedCount < CAT_COUNT_MIN) expectedCount = CAT_COUNT_MIN;
    if (expectedCount > CAT_COUNT_MAX) expectedCount = CAT_COUNT_MAX;
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == expectedCount);

    int seen = 0;
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Cat) continue;
        const double margin = CAT_BODY_RADIUS + 8.0;
        const double expectedX = prng_uniform(side, margin, 1920.0 - margin);
        const double expectedSpeed = prng_uniform(side, CAT_WALK_SPEED_MIN, CAT_WALK_SPEED_MAX);
        const double dirCoin = prng_uniform(side, 0.0, 1.0);
        const double expectedDir = (dirCoin < 0.5) ? -1.0 : 1.0;
        const uint32_t expectedSeed = prng_next_u32(side);
        const double expectedTimer = prng_uniform(side, CAT_WALK_DURATION_MIN, CAT_WALK_DURATION_MAX);
        const uint8_t expectedNameIndex = static_cast<uint8_t>(prng_index(side,
            static_cast<uint32_t>(sizeof(CAT_NAME_POOL) / sizeof(CAT_NAME_POOL[0]))));
        const uint8_t expectedCoatVariantIndex = static_cast<uint8_t>(prng_index(side,
            static_cast<uint32_t>(CAT_COAT_VARIANT_COUNT)));

        Assert::IsTrue(e.x == Near(expectedX));
        Assert::IsTrue(e.vx == Near(expectedSpeed * expectedDir));
        Assert::IsTrue(e.seed == expectedSeed);
        Assert::IsTrue(e.stateTimer == Near(expectedTimer));
        Assert::IsTrue(e.nameIndex == expectedNameIndex);
        Assert::IsTrue(e.coatVariantIndex == expectedCoatVariantIndex);
        ++seen;
    }
    Assert::IsTrue(seen == expectedCount);
}

TEST_METHOD(SimSetCritterNoneClearsAmbientCats) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) >= CAT_COUNT_MIN);

    sim_set_critter(sim, CritterKind::None);
    Assert::IsTrue(sim.currentCritter == CritterKind::None);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 0);
    Assert::IsTrue(count_kind(sim, EntityKind::Bunny) == 0);
}

TEST_METHOD(SwitchingBetweenCritterSpeciesReplacesThePreviousSpecies) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) >= CAT_COUNT_MIN);
    Assert::IsTrue(count_kind(sim, EntityKind::Sheep) == 0);

    sim_set_critter(sim, CritterKind::Sheep);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 0);
    Assert::IsTrue(count_kind(sim, EntityKind::Sheep) >= SHEEP_COUNT_MIN);

    sim_set_critter(sim, CritterKind::Cat);
    Assert::IsTrue(count_kind(sim, EntityKind::Sheep) == 0);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) >= CAT_COUNT_MIN);
}

TEST_METHOD(SimSetSceneGatesActiveCatToGrass) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    const int catsGrass = count_kind(sim, EntityKind::Cat);
    Assert::IsTrue(catsGrass >= CAT_COUNT_MIN);

    sim_set_scene(sim, Scene::Desert);
    Assert::IsTrue(sim.currentCritter == CritterKind::Cat);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 0);

    sim_set_scene(sim, Scene::Winter);
    Assert::IsTrue(sim.currentCritter == CritterKind::Cat);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 0);

    sim_set_scene(sim, Scene::Grass);
    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == catsGrass);
}

TEST_METHOD(ClickWithinCATPOUNCERADIUSPouncesTowardTheClick) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    keep_first_cat_only(sim);

    Entity& cat = sim.entities.front();
    cat.x = 500.0;
    cat.vx = -CAT_WALK_SPEED_MIN;
    cat.age = 5.0;

    sim_apply_click(sim, click_event(cat.x + 16.0, sim.windowHeight - 20.0));

    const Entity& after = sim.entities.front();
    Assert::IsTrue(after.state == CAT_STATE_POUNCING);
    Assert::IsTrue(after.stateTimer == Near(CAT_POUNCE_DURATION));
    Assert::IsTrue(after.age == Near(0.0));
    Assert::IsTrue(after.vx == Near(CAT_POUNCE_SPEED));
}

TEST_METHOD(ClickOutsideCATPOUNCERADIUSLeavesCatAlone) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    keep_first_cat_only(sim);

    Entity& cat = sim.entities.front();
    cat.x = 500.0;
    cat.vx = -CAT_WALK_SPEED_MIN;
    const uint8_t stateBefore = cat.state;
    const double vxBefore = cat.vx;

    sim_apply_click(sim, click_event(cat.x + CAT_POUNCE_RADIUS + 5.0, sim.windowHeight - 20.0));

    const Entity& after = sim.entities.front();
    Assert::IsTrue(after.state == stateBefore);
    Assert::IsTrue(after.vx == Near(vxBefore));
}

TEST_METHOD(CatsDoNotGreetOtherCats) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    keep_first_cat_only(sim);

    Entity first = sim.entities.front();
    first.x = 400.0;
    first.vx = CAT_WALK_SPEED_MIN;
    first.state = CAT_STATE_WALKING;
    first.stateTimer = 10.0;
    first.age = SHEEP_GREET_MIN_AGE + 1.0;
    Entity second = first;
    second.x = first.x + 20.0;
    second.vx = -CAT_WALK_SPEED_MIN;
    sim.entities.clear();
    sim.entities.push_back(first);
    sim.entities.push_back(second);

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 2);
    for (const Entity& e : sim.entities) {
        if (e.kind == EntityKind::Cat) Assert::IsTrue(e.state != SHEEP_STATE_GREETING);
    }
}

TEST_METHOD(CatsDoNotGreetSheep) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Cat);
    keep_first_cat_only(sim);

    Entity cat = sim.entities.front();
    cat.x = 400.0;
    cat.vx = CAT_WALK_SPEED_MIN;
    cat.state = CAT_STATE_WALKING;
    cat.stateTimer = 10.0;
    cat.age = SHEEP_GREET_MIN_AGE + 1.0;

    Entity sheep{};
    sheep.kind = EntityKind::Sheep;
    sheep.size = SHEEP_BODY_RADIUS;
    sheep.x = cat.x + 20.0;
    sheep.y = sim.windowHeight - SHEEP_BODY_HEIGHT - SHEEP_LEG_LENGTH;
    sheep.vx = -SHEEP_WALK_SPEED_MIN;
    sheep.age = SHEEP_GREET_MIN_AGE + 1.0;
    sheep.lifetime = -1.0;
    sheep.state = SHEEP_STATE_WALKING;
    sheep.stateTimer = 10.0;

    sim.entities.clear();
    sim.entities.push_back(cat);
    sim.entities.push_back(sheep);

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(count_kind(sim, EntityKind::Cat) == 1);
    Assert::IsTrue(count_kind(sim, EntityKind::Sheep) == 1);
    for (const Entity& e : sim.entities) Assert::IsTrue(e.state != SHEEP_STATE_GREETING);
}
};
}
